File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    answer = 0
    total_topping_count = len(set(topping))
    total_topping = {}
    
    for item in topping:
        if item in total_topping:
            total_topping[item] += 1
        else:
            total_topping[item] = 1
        
    check_topping = set()
    for i in range(len(topping)):
        current_item = topping[i]
        total_topping[current_item] -= 1
        check_topping.add(current_item)
        
        if total_topping[current_item] == 0:
            total_topping_count -= 1
        
        if total_topping_count == len(check_topping):
            answer += 1
    
    return answer

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let answer = 0;
    const n = topping.length;
    
    const ref = new Map();
    topping.forEach((item) => {
        ref.set(item, ref.has(item) ? ref.get(item) + 1 : 1);
    });
    
    let totalCount = [...ref.keys()].length;
    const checkSet = new Set();
    topping.forEach((item) => {
        checkSet.add(item);
        if (ref.has(item) && ref.get(item) > 0) ref.set(item, ref.get(item) - 1);
        
        if (ref.get(item) === 0) {
            ref.delete(item);
            totalCount--;
        }
        
        if (checkSet.size === totalCount) answer++;
    })
    
    return answer;
}